### PR TITLE
Update previous epoch calculation

### LINKED ISSUE #382: Make `get_previous_epoch` return `current_epoch - 1` in all cases
### What is wrong?
Part of Trinity #379
Spec PRs:
   - Merged PR: ethereum/eth2.0-specs#711

### How can it be fixed
Update `BeaconState.previous_epoch()` function

### PR DESCRIPTION
**Pull request by ralexstokes**
### What was wrong?

Fixes #382.

Replaces #398.

### How was it fixed?

Started rebasing #398 and realized we want to handle some of the tests differently.

This has the same content but keeps some of the tests... after further inspection I think what I had in #398 was actually weakening the test coverage.

[//]: # (For important changes, please add a new entry to the release notes file)
[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by ralexstokes** [comment]
@ChihChengLiang @hwwhww @NIC619 given that you reviewed the other PR, i'll tag you here.

like stated above, i started rebasing and realized we want to handle some of the testing in a better way.

the one thing i could call special attention to is `test_justification_without_mock`. it seems like it should yield a bitfield `0b0` _unless_ we provide attestations. and it doesn't seem like we will get `0b11` until we run through two epoch processings. so i think that the prior implementation was incorrect and has now been fixed (although is not testing the scenario where we do in fact hit `0b11`) and that there was a bug in the previous implementation of `_current_previous_epochs_justifiable`. if i'm wrong, then there is a bug in the new `_is_epoch_justifiable` and will happily track it down so let me know :)

thanks!

**Comment by ChihChengLiang** [comment]
@ralexstokes Feel free to fix them, those functions are implemented before v0.2.0. Prior to this PR, previous epoch and current epoch is 0 on the genesis, that's the very reason why `test_justification_without_mock` is getting 0b11. `_current_previous_epochs_justifiable` is also an awkward move to address the behavior of previous epoch and current epoch near genesis. As the consistent behavior introduced, it is safe to nuke `_current_previous_epochs_justifiable` and replace it with `_is_epoch_justifiable`.

**Comment by ralexstokes** [comment]
@ChihChengLiang great -- went ahead and did that! just to confirm, we do want this w/ the latest implementation correct? https://github.com/ethereum/trinity/pull/469/files#diff-9ffda674671b43bc2ac2193ad772c6a5R319

this is my understanding but we want to be extra careful about justifying the correct stuff :)

File: eth2/beacon/epoch_processing_helpers.py [hunk 1]
```diff
@@ -145,69 +145,55 @@ def get_attestations_for(root: Hash32) -> Sequence[PendingAttestationRecord]:
...
-        previous_epoch_boundary_attester_indices,
-        config.MAX_DEPOSIT_AMOUNT,
+@to_tuple
+def _get_epoch_boundary_attesting_indices(state: 'BeaconState',
+                                          attestations: Sequence[PendingAttestationRecord],
+                                          epoch: Epoch,
+                                          config: Eth2Config) -> Iterable[ValidatorIndex]:
+    relevant_attestations = (
+        a for a in attestations
+        if a.data.epoch_boundary_root == get_block_root(
```
Comment (ChihChengLiang) [inline eth2/beacon/epoch_processing_helpers.py]: This `get_block_root` is called with same inputs many times in the loop

File: eth2/beacon/epoch_processing_helpers.py [hunk 2]
```diff
@@ -145,69 +145,55 @@ def get_attestations_for(root: Hash32) -> Sequence[PendingAttestationRecord]:
...
-        state.previous_justified_epoch,
-        previous_epoch_boundary_root,
-        CommitteeConfig(config),
-    )
 
-    previous_epoch_boundary_attesting_balance = get_total_balance(
-        state.validator_balances,
-        previous_epoch_boundary_attester_indices,
-        config.MAX_DEPOSIT_AMOUNT,
+@to_tuple
```
Comment (ChihChengLiang) [inline eth2/beacon/epoch_processing_helpers.py]: No need for to_tuple if we are returning tuple already

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 3]
```diff
@@ -128,47 +129,31 @@ def process_eth1_data_votes(state: BeaconState, config: Eth2Config) -> BeaconSta
...
-    previous_epoch_active_validator_indices = get_active_validator_indices(
+    active_validator_indices = get_active_validator_indices(
         state.validator_registry,
-        previous_epoch,
-    )
-    current_total_balance = get_total_balance(
-        state.validator_balances,
-        current_epoch_active_validator_indices,
-        config.MAX_DEPOSIT_AMOUNT,
+        epoch
```
Comment (ChihChengLiang) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]:
```
epoch,
```

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 4]
```diff
@@ -209,15 +194,20 @@ def _get_finalized_epoch(
...
     current_epoch = state.current_epoch(config.SLOTS_PER_EPOCH)
-    previous_epoch = state.previous_epoch(config.SLOTS_PER_EPOCH, config.GENESIS_EPOCH)
+    previous_epoch = state.previous_epoch(config.SLOTS_PER_EPOCH)
 
-    current_epoch_justifiable, previous_epoch_justifiable = _current_previous_epochs_justifiable(
+    current_epoch_justifiable = _is_epoch_justifiable(
         state,
+        state.current_epoch_attestations,
         current_epoch,
+        config
```
Comment (ChihChengLiang) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]:
```
config,
```

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 5]
```diff
@@ -209,15 +194,20 @@ def _get_finalized_epoch(
...
+    )
+    previous_epoch_justifiable = _is_epoch_justifiable(
+        state,
+        state.previous_epoch_attestations,
         previous_epoch,
-        config,
+        config
```
Comment (ChihChengLiang) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]:
```
config,
```
FYI https://github.com/ethereum/snake-charmers-tactical-manual/blob/master/style-guide.md#dangling-commas

File: tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py [hunk 6]
```diff
@@ -217,7 +217,7 @@ def test_only_process_eth1_data_votes_per_period(sample_beacon_state_params, con
         ),
     )
 )
-def test_current_previous_epochs_justifiable(
+def test_is_epoch_justifiable(
```
Comment (ChihChengLiang) [inline tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py]: We can get rid of `current_epoch` and `previous_epoch` in this test actually. This test was intended to test `_current_previous_epochs_justifiable` but now `_is_epoch_justifiable`, and only the former needs them both.

**Comment by ChihChengLiang** [comment]
Great work! Only leave some nitpick here. Justification and finalization look good to me, their core logic did not change much in the spec and the test cases are still valid.

File: eth2/beacon/epoch_processing_helpers.py (hunk 1 above)
Comment (ralexstokes) [reply eth2/beacon/epoch_processing_helpers.py]: good catch

File: eth2/beacon/epoch_processing_helpers.py (hunk 2 above)
Comment (ralexstokes) [reply eth2/beacon/epoch_processing_helpers.py]: another good catch

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py (hunk 5 above)
Comment (ralexstokes) [reply eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: yep, totally

File: tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py (hunk 6 above)
Comment (ralexstokes) [reply tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py]: 👍

**Pull request merged**